src/ORM/DataList.php

```diff
@@ -308,72 +313,118 @@ public function distinct($value)
+            }
+            if (is_null($sort)) {
+                // Set an an empty array here to cause any existing sort on the DataLists to be wiped
+                // later on in this method
+                $sort = [];
+            } elseif (empty($sort)) {
+                throw new InvalidArgumentException('Invalid sort paramater');
+            }
+            // If $sort is string then convert string to array to allow for validation
+            if (is_string($sort)) {
+                $newSort = [];
+                // Making the assumption here there are no commas in column names
+                // Other parts of silverstripe will break if there are commas in column names
+                foreach (explode(',', $sort) as $colDir) {
+                    // Using regex instead of explode(' ') in case column name includes spaces
+                    if (preg_match('/^(.+) ([^"]+)$/i', trim($colDir), $matches)) {
+                        list($column, $direction) = [$matches[1], $matches[2]];
+                    } else {
+                        list($column, $direction) = [$colDir, 'ASC'];
+                    }
+                    $newSort[$column] = $direction;
+                }
+                $sort = $newSort;
+            }
+        }
+        // Validate that column names are not raw sql
+        if (is_array($sort)) {
+            foreach ($sort as $column => $direction) {
+                $this->validateColumnHasNoRawSql($column);
+                $this->validateSortDirection($direction);
+            }
+        }
+        return $this->alterDataQuery(function (DataQuery $query, DataList $list) use ($sort) {
+            // Wipe the sort
+            $query->sort(null, null);
+            foreach ($sort as $column => $direction) {
+                $list->applyRelation($column, $relationColumn, true);
+                $query->sort($relationColumn, $direction, false);
             }
+        });
+    }
 
-            $sort = [$col => $dir];
-        } else {
-            $sort = func_get_arg(0);
+    private function validateColumnHasNoRawSql(string $column): void
+    {
+        // Remove quotes around the column name
+        $col = trim(trim($column), '"');
+        // Using regex because $this->canSortBy() failed on complex relations e.g. Team.Founder.Surname and
+        // also increased the number of db queries
+        // Search for non-alphanumeric characters in the column name rather than ^[^a-zA-Z0-9_]$ to account
+        // for column names that include macrons etc
+        if (preg_match('#[()[\]{}+\-*/=:?\\;|&\'"`<>^%,\s]#', $col)) {
+            $message = implode(' ', [
+                "Column name $column includes characters that could lead to SQL injection vulnerabilities.",
+                'Rename this column or use DataList::orderBy() for sorting instead.'
+            ]);
+            throw new InvalidArgumentException($message);
         }
```

Comment:
I'm still on the fence about doing this with regex.... I'd like a third opinion on it before merging. I _think_ I'm okay with it on the basis that this is a major release (so if it breaks something for someone it's not the end of the world) and it's _relatively_ unlikely people are using these characters in their column names...

@silverstripe/core-team @maxime-rainville do any of you have opinions on this?
See also discussion on https://github.com/silverstripe/silverstripe-framework/pull/10600#discussion_r1039052177 and https://github.com/silverstripe/silverstripe-framework/pull/10600#discussion_r1039051614

Reply:
In fact the more I think about this, the less okay with it I am, for the same reasons as are being discussed in https://github.com/silverstripe/silverstripe-framework/issues/10605 - if we aren't restricting db columns to some defined list of characters globally, then this regex effectively restricts where and how some db columns can be used for sorting.

For example, gridfield sort columns currently use `sort()`, not `orderBy()` - so otherwise valid columns simply can't be sorted in gridfield columns if we implement this as is.

Reply:
> gridfield sort columns currently use sort(), not orderBy()

Standard so far is to just change the ->sort() call in GridField to ->orderBy() - it's functionally the same.

Reply:
Don't really like the approach here. I'm not necessarily against the idea of explicitly excluding some of those special characters from column names. However, it seems to me the proper way of handling this would be to have a dedicated approach to make sure whatever we put in the column name always gets properly escape when we generate the query.

My instinct would be to look at how `filter` handles the column name escaping and try to replicate that as much as possible. This seems to be mostly handle by [`SearchFilter::getDbName()`](https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/Filters/SearchFilter.php#L258)

Reply:
`Convert::raw2sql()` may be appropriate as well. This is what `DataObjectSchema::tableName()` uses

Reply:
Apologies in advance if I get anything wrong here, this is a confusing topic

It's important to note there's a big different between 2 seemingly similar but totally different query syntax's
a) Silverstripe relations e.g. `'MyModel.MyField'`
b) SQL relations e.g. `'"MyTable"."MyField"'`

Silverstripe relations will work correctly in the new sort().  SQL relations are raw sql and require the new orderBy().  


```php
sort('MyModel.MyColumn'); // silverstripe relation
sort('MyModel.MyRelation.MyColumn'); // silverstripe relation
sort('"MyTableName"."MyColumn"'); // this is raw sql - throw an exception
```

Reply:
Have refactored out the regex